Added Documentation on Elastic Inference and MMS (#715)
awslabs/mxnet-model-server: docs/elastic_inference.md

# Model Serving with Amazon Elastic Inference 

## Contents of this Document
* [Introduction](#introduction)
* [Custom Service](#custom-service)
* [Creating a EC2 instance with EIA support](#creating-a-ec2-instance-with-eia-support)
* [Custom Service file with EIA](#custom-service-file-with-eia)
* [Running elastic inference on a resnet-152](#running-elastic-inference-on-a-resnet-152)
## Introduction

> Amazon Elastic Inference allows you to attach low-cost GPU-powered acceleration to Amazon EC2 and Amazon SageMaker instances to reduce the cost of running deep learning inference by up to 75%. With MMS it is easy to deploy an MXNet based model, taking advantage of the attacble hardware accelerator.

Reviewer: ```suggestion
Amazon Elastic Inference allows you to attach low-cost GPU-powered acceleration to Amazon EC2 and Amazon SageMaker instances to reduce the cost of running deep learning inference by up to 75%. With MMS it is easy to deploy an MXNet based model, taking advantage of the attachable hardware accelerator.
```

> In this document we explore using Elastic Inference Acclerator (EIA) attached to a Compute Optimized EC2 instance.

Reviewer: Can you add something to clarify how EI isn't EIA? In the previous paragraph you introduce the accelerator, so just carry on with that.

## Custom Service

The capability to run model inference with the Elastic Inference Accelerator (EIA) can be achieved by building the custom service to use the EIA context rather than a GPU or CPU context. An MXNet version with support for EIA is required.

To understand the basics of writing a custom service file refer to the [Custom Service Documentation](https://github.com/awslabs/mxnet-model-server/blob/master/docs/custom_service.md).

## Creating a EC2 instance with EIA support 

To Create an EC2 instance with EIA support there are few pre-requisites. These include:

> 1. Configuring a Secuirty Group

Reviewer 2: ```suggestion
 1. Configuring a Security Group
```

Reviewer: You could provide the direct links to these steps like I've done here in the [DLAMI IE prereq docs](https://docs.aws.amazon.com/dlami/latest/devguide/ei-prerequisites.html).

2. Creating a secure Endpoint.
3. Creating a IAM role policy.
    
The above steps are explored in detail in [AWS Elastic Inference documentation](https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/working-with-ei.html)  

On completing the above steps, following steps need to be followed in launching a instance with EIA support

1. Open the Amazon EC2 console at https://console.aws.amazon.com/ec2/.

2. Choose Launch Instance.

> 3. Under Choose one of the Deep Learning AMIs, we recommend Deep Learning AMI v20. This is required to use MXNet with EIA.

Reviewer 2: Why not say v20 or later? DL AMI releases a new version every month or so, and you will probably not update this page every DL AMI release.

4. Under Choose an Instance Type, we recommend a compute optimized EC2 instance such as c5.2xlarge.

5. Choose Next: Configure Instance Details.

6. Under Configure Instance Details, check the configuration settings. Ensure that you are using the VPC with the security groups for the instance and the Amazon EI accelerator that was setup earlier. For more information, see Configuring Your Security Groups for Amazon EI.

7. For IAM role, select the role that you created in the Configuring an Instance Role with an Amazon EI Policy procedure explained in the above documentation.

8. Select Add an Amazon EI accelerator.

> 9. Select the size of the Amazon EI accelerator. Your options are: eia1.medium, eia1.large, and eia1.xlarge. 

Reviewer: It might be a good idea to mention that certain models require a larger EIA. Not sure if there are guidelines for this yet.

Author: I am not sure of this either, However I will add a brief note on the same.

10. (Optional) You can choose to add storage and tags by choosing Next at the bottom of the page. Or, you can let the instance wizard complete the remaining configuration steps for you.

11. Review the configuration of your instance and choose Launch.

12. You are prompted to choose an existing key pair for your instance or to create a new key pair. 

```
WARNING

>     Don’t select the Proceed without a key pair option. If you launch your instance without a key pair, then you can’t connect to it.

Reviewer: Using a code block makes this not wrap, and important info isn't easily read on the screen. I'd take it out of the code block and add bold to it...

    After making your key pair selection, choose Launch Instances.
```

It can take a few minutes for the instance to be ready so that you can connect to it. Check that your instance has passed its status checks. You can view this information in the Status Checks column.

## Custom Service file with EIA 

> A custom service, for MXNet that supports EIA, consists of two contexts. The input data would be loaded from cpu context and the symbol is bound to eia context, i.e:

Reviewer: ```suggestion
You use two different processing contexts with MXNet and EIA.
```

> ```text

Reviewer: I don't think you want to put important info in a code block like this.
```suggestion

```

1. mxnet.cpu() - used for loading up input data

> 2. mxnet.eia() - Used for binding network symbols and params to an attached EIA instance.

Reviewer: ```suggestion
2. mxnet.eia() - used for binding network symbols and params to an attached EIA instance
```

> ```

Reviewer: ```suggestion

```

        # Symbol Context set to eia
        self.mxnet_ctx = mx.eia()
        self.data_ctx = mx.cpu()
        sym, arg_params, aux_params = mx.model.load_checkpoint(checkpoint_prefix, self.epoch)

        # noinspection PyTypeChecker
        self.mx_model = mx.mod.Module(symbol=sym, context=self.mxnet_ctx,
                                      data_names=data_names, label_names=None)
        self.mx_model.bind(for_training=False, data_shapes=data_shapes)
        self.mx_model.set_params(arg_params, aux_params, allow_missing=True, allow_extra=True)
    
    def inference(self, model_input):
        # NOT COMPLETE CODE, refer template above for it.
        .....
        .....
        .....     
        model_input = [item.as_in_context(self.data_ctx) for item in model_input]

```

The above code shows initialization of two contexts, one for data (on CPU) and other for symbols (on EIA).

Once we have the code ready. We can build a model archive consumable by MMS, using the model-archiver. The [model-archiver](https://github.com/awslabs/mxnet-model-server/blob/master/model-archiver/README.md) tool enables the create to an archive understood by MMS.

```python

> model-archiver --model-name <model-name> --handler model_service:handle --export-path <output-dir> --model-path <model_dir> --runtime python

Reviewer 3: Why -- showing red here?

Reviewer: Maybe they're not really dashes, and they're some unicode variant? Try deleting them and typing it again.

```

This will create file ```<model-name>.mar``` in the directory ```<output-dir>```.

## Running elastic inference on a resnet-152

> A prebuilt Amazon Elastic Inference supporting resnet-152  model archive can be downloaded using the following command:

Reviewer: ```suggestion
A pre-built ResNet-152 model archive that uses Amazon Elastic Inference can be downloaded using the following command:
```

> ```bash

Reviewer 3: I think we should merge all these steps into a single start cli:
mxnet-mode-server --start --models reset-152-eai=https://s3.amazonaws.com/model-server/model_archive_1.0/resnet-152-eia.mar

$ wget https://s3.amazonaws.com/model-server/model_archive_1.0/resnet-152-eia.mar
```

> **NOTE:** The above archive will exclusively only work with EIA enabled instances, of EC2 and DLAMI v20.

Reviewer: ```suggestion
**NOTE:** The above archive will only work on EIA-enabled instances.
```

> Start the EC2 instance, with DLAMI v20There are two conda environments with mxnet elastic inference and mxnet-model-server pre-installed. Either of the two can be used.

Reviewer: ```suggestion
Start the EIA-enabled EC2 instance. If using a Deep Learning AMI, there are two Conda environments (one for Python 2 and one for Python 3), both of which come with MXNet built will EI support and MXNet Model Server. Either of the two can be used.
```

```bash
# python 3 
$ source activate amazonei_mxnet_p36

# python 2 
$ source activate amazonei_mxnet_p27
```

> After entering the virtual environment. We start MMS.

Reviewer: ```suggestion
After entering the one of the Conda environments, we start MMS.
```

```bash
# Start MMS
$ mxnet-model-server --start
```

> We register the Resnet-152 model and start a single worker for it

Reviewer: ```suggestion
We register the ResNet-152 model and start a single worker for it:
```

```bash

> # Register resent-18 model archive with Amazon EI support

Reviewer: ```suggestion
# Register the ResNet-152 model archive with Amazon EI support
```

$ curl -X POST 127.0.0.1:8081/models?url=https://s3.amazonaws.com/model-server/model_archive_1.0/resnet-152-eia.mar

# Start a single worker for the registered model

> 

Reviewer: ```suggestion
```

$ curl -X PUT '127.0.0.1:8081/models/resnet-152?min_worker=1&synchronous=true'

```

> Now the model is ready for some inference requests. Let us download a kitten image for classification.

Reviewer: You should mention how the port is different... It's easy to miss and not understand what is happening.

```bash
$ curl -O https://s3.amazonaws.com/model-server/inputs/kitten.jpg
$ curl -X POST http://127.0.0.1:8080/predictions/resnet-152 -T kitten.jpg
```

The predict endpoint will return a prediction response in JSON. It will look something like the following result:

```json
[
  {
    "probability": 0.7148934602737427,
    "class": "n02123045 tabby, tabby cat"
  },
  {
    "probability": 0.22877734899520874,
    "class": "n02123159 tiger cat"
  },
  {
    "probability": 0.04032360762357712,
    "class": "n02124075 Egyptian cat"
  },
  {
    "probability": 0.008370809257030487,
    "class": "n02127052 lynx, catamount"
  },
  {
    "probability": 0.0006728142034262419,
    "class": "n02129604 tiger, Panthera tigris"
  }

> ```

Reviewer 3: This looks strange, they are all red

Reviewer: I imagine that if get rid of the `text` code block, you might find things render better...

> Resnet-152 identified the tabby cat using Elastic Inference Accelerator while being hosted on MMS. Serving on Amazon EI instances reduces, the cost while offering performance of a GPU for inference tasks   .

Reviewer: ```suggestion
ResNet-152 identified the tabby cat using Elastic Inference Accelerator while being hosted on MMS. Serving on Amazon EI instances reduces inference costs while benefiting from the performance of a GPU for inference tasks.
```
